### PR TITLE
Checklist models

### PR DESCRIPTION
**Pull request by paopow**
Update the Todo model and add the ChecklistTemplate model.

### CODE REVIEW

File: orchestra/models/core/models.py [hunk 1]
```diff
@@ -607,6 +646,12 @@ class Meta:
     completed = models.BooleanField(default=False)
     start_by_datetime = models.DateTimeField(null=True, blank=True)
     due_datetime = models.DateTimeField(null=True, blank=True)
+    skipped = models.BooleanField(default=False)
+    parent_todo = models.ForeignKey(
+        'self', null=True, related_name='parent')
+    template = models.ForeignKey(
```
Comment (marcua) [inline orchestra/models/core/models.py:652]: whenever null=True, also do blank=True so you can save it in admin :)

File: orchestra/models/core/models.py [hunk 2]
```diff
@@ -607,6 +646,12 @@ class Meta:
     completed = models.BooleanField(default=False)
     start_by_datetime = models.DateTimeField(null=True, blank=True)
     due_datetime = models.DateTimeField(null=True, blank=True)
+    skipped = models.BooleanField(default=False)
```
Comment (marcua) [inline orchestra/models/core/models.py:649]: skipped_datetime so we can reconstruct when these happened :)

File: orchestra/models/core/models.py [hunk 3]
```diff
@@ -607,6 +646,12 @@ class Meta:
     completed = models.BooleanField(default=False)
```
Comment (marcua) [inline orchestra/models/core/models.py:646]: we could consider completed_datetime, but that could be a separate migration

File: orchestra/models/core/models.py [hunk 4]
```diff
@@ -607,6 +646,12 @@ class Meta:
...
     start_by_datetime = models.DateTimeField(null=True, blank=True)
     due_datetime = models.DateTimeField(null=True, blank=True)
+    skipped = models.BooleanField(default=False)
+    parent_todo = models.ForeignKey(
+        'self', null=True, related_name='parent')
+    template = models.ForeignKey(
+        ChecklistTemplate, null=True, related_name='template')
+    log = JSONField(default={'actions': []})
```
Comment (marcua) [inline orchestra/models/core/models.py:654]: activity_log ?

File: orchestra/models/core/models.py [hunk 5]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
     end_date = models.DateField(null=True, blank=True)
 
 
+class ChecklistTemplate (ChecklistTemplateMixin, BaseModel):
```
Comment (marcua) [inline orchestra/models/core/models.py:580]: Given the objects are called Todos, can we call this TodoListTemplate and TotoListTemplateMixin?

File: orchestra/models/core/models.py [hunk 6]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+    """
+    A checklist template
+
+    Attributes:
+        slug (str):
+            The string that use to refer the checklist
```
Comment (marcua) [inline orchestra/models/core/models.py:586]: A unique key for the todo list

File: orchestra/models/core/models.py [hunk 7]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+        name (str):
+            The name of the checklist
```
Comment (marcua) [inline orchestra/models/core/models.py:588]: A human-readable name for the todo list

File: orchestra/models/core/models.py [hunk 8]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+        description (str):
+            A text description of the checklist
```
Comment (marcua) [inline orchestra/models/core/models.py:590]: text -> longer

File: orchestra/models/core/models.py [hunk 9]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+        creator (orchestra.models.Worker)
+            The worker who creates the checklist. This field is null
```
Comment (marcua) [inline orchestra/models/core/models.py:592]: creates -> created

File: orchestra/models/core/models.py [hunk 10]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+    """
+    class Meta:
+        app_label = 'orchestra'
+
+    slug = models.CharField(max_length=200, unique=True)
+    name = models.CharField(max_length=200)
+    description = models.TextField()
+    creator = models.ForeignKey(
+        Worker, null=True, related_name='creator', on_delete=models.CASCADE)
+    todos = JSONField(default={'list': []})
```
Comment (marcua) [inline orchestra/models/core/models.py:605]: document the format

File: orchestra/models/core/models.py [hunk 11]
```diff
@@ -576,6 +577,34 @@ class PayRate(PayRateMixin, models.Model):
...
+            A JSON blob that describe the todos in the checklist
+    """
+    class Meta:
+        app_label = 'orchestra'
+
+    slug = models.CharField(max_length=200, unique=True)
+    name = models.CharField(max_length=200)
+    description = models.TextField()
+    creator = models.ForeignKey(
+        Worker, null=True, related_name='creator', on_delete=models.CASCADE)
```
Comment (marcua) [inline orchestra/models/core/models.py:604]: also blank=True

Comment (marcua) [inline orchestra/models/core/models.py:604]: should the delete SET_NULL so we don't lose the template?

File: orchestra/models/core/models.py (hunk 2 above)
Comment (adbharadwaj) [reply orchestra/models/core/models.py:649]: @marcua A few doubts. :)
- Arent we planning to use activity log to reconstruct what happened?
- What happens when a user skips and then adds back? I suppose we set the field to null when someone unskips a todo.

File: orchestra/models/core/models.py (hunk 4 above)
Comment (adbharadwaj) [reply orchestra/models/core/models.py:654]: I like `activity_log`.

File: orchestra/models/core/models.py (hunk 11 above)
Comment (adbharadwaj) [reply orchestra/models/core/models.py:604]: @marcua Shouldn't manage different versions of a template (add a version column?). I suppose we are going to update templates over time.

Comment (marcua) [reply orchestra/models/core/models.py:604]: I'm pro-versioning, though I think it's too early for the needs of the product right now. I still think we should SET_NULL to be safe for the time being.

File: orchestra/models/core/models.py (hunk 2 above)
Comment (marcua) [reply orchestra/models/core/models.py:649]: In general, the best solution for activity logging is to not write it out to the DB, but instead to event log into tools like Snowplow or Google Analytics. So I'm a little skeptical of the activity_log in the long-term, and am a big fan of skipped_datetime as an imperfect signal.

Setting to null if unskipped seems great!

Comment (adbharadwaj) [reply orchestra/models/core/models.py:649]: Cool :) I am a big fan of Google analytics event tags too :)

**Pull request merged**